src/dmx/DMX.cpp

```diff
@@ -0,0 +1,134 @@
+
+#include <iostream>
+#include <math.h>
+#include <mutex>
+#include <stdlib.h>
+#include <thread>
+#include <unistd.h>
+
+#include <ola/DmxBuffer.h>
+#include <ola/Logging.h>
+#include <ola/client/StreamingClient.h>
+
+#include "DMX.hpp"
+
+#include "stdlib.h"
+#include "stdio.h"
+#include "string.h"
+
+using namespace std;
+
+/*
+  This module speaks DMX to spotlights via Open Lighting https://wiki.openlighting.org/
+
+ */
+
+
+void DMX::init() {
+    std::cout << "DMX: init()" << std::endl;
+    // DMX "universe" to use for sending data.
+    // In OLA, this is configured with ola_patch
+    universe = 0;
+
+    // turn on OLA logging
+    ola::InitLogging(ola::OLA_LOG_WARN, ola::OLA_LOG_STDERR);
+    blackout_buffer.Blackout(); // Set all channels to 0
+
+    // Setup the client; this connects to the OLA server
+    if (!ola_client.Setup()) {
+        std::cerr << "OLA DMX Setup failed" << endl;
+        exit(1);
+    }
+
+    // Black the lights out
+    blackout_buffer.SetChannel(0, 0);
+    send_dmx(blackout_buffer);
+}
+
+DMX::DMX() : m_led(0,0,0),
+	     ola_client((ola::client::StreamingClient::Options())) {}
+
+void DMXSerial::send(unsigned char const *buf, size_t const len) {
+    std::cout << "DMX: send()" << std::endl;
+    size_t buf_len(len);
+    const size_t MAX_VALUES_USED = 4; // r g b w
+
+    if (len > MAX_VALUES_USED) {
+	cerr << "Hey! More than one light specified for the cross spotlights, but I'm only using the first one !" << endl;
+	for(int i =0; i<len; i++) {
+	    std::cout << (int) buf[i] << " ";
+	}
+	std::cout << endl;
+    }
+
+
+    if (len < MAX_VALUES_USED) {
+	cerr << "Hey! I need at least " << MAX_VALUES_USED << " data bytes, but I only got " << len << " bytes!! Exiting DMX::send." << endl;
+	return;
+    }
+
+    // ----- begin buffer changed check
+    // most of the time there is no change to buf
+    // this check avoids calling the update machinery if there has been no change
+
+    if (len > m_last_buf_len) {
+        m_last_buf = static_cast<unsigned char *>(realloc(m_last_buf, len));
+        m_last_buf_len = len;
+    }
+
+    bool buf_changed = false;
+    for (int i = 0; i < MAX_VALUES_USED; i++) {
+        if (m_last_buf[i] != buf[i]) {
+            buf_changed = true;
+            break;
+        }
+    }
+
+    if (buf_changed) {
+        memcpy(m_last_buf, buf, len);
+    } else {
+        return;
+    }
+
```

Comment:
Yes. I may remove this code, because it doesn't matter how  often you write to the DMX bus, and it seems unlikely OSC will overwhelm the DMX bus.